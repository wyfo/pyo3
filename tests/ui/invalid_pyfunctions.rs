use pyo3::{coroutine::CancelHandle, prelude::*};

#[pyfunction]
fn generic_function<T>(value: T) {}

#[pyfunction]
fn impl_trait_function(impl_trait: impl AsRef<PyAny>) {}

#[pyfunction]
fn wildcard_argument(_: i32) {}

#[pyfunction]
fn destructured_argument((a, b): (i32, i32)) {}

#[pyfunction]
fn function_with_required_after_option(_opt: Option<i32>, _x: i32) {}

#[pyfunction(cancel_handle = "cancel")]
fn sync_fn_with_cancel_handle(cancel: CancelHandle) {}

#[pyfunction(cancel_handle = "cancel")]
async fn missing_cancel_handle() {}

fn main() {}
