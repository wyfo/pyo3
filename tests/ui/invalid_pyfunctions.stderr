error: Python functions cannot have generic type parameters
 --> tests/ui/invalid_pyfunctions.rs:4:21
  |
4 | fn generic_function<T>(value: T) {}
  |                     ^

error: Python functions cannot have `impl Trait` arguments
 --> tests/ui/invalid_pyfunctions.rs:7:36
  |
7 | fn impl_trait_function(impl_trait: impl AsRef<PyAny>) {}
  |                                    ^^^^

error: wildcard argument names are not supported
  --> tests/ui/invalid_pyfunctions.rs:10:22
   |
10 | fn wildcard_argument(_: i32) {}
   |                      ^

error: destructuring in arguments is not supported
  --> tests/ui/invalid_pyfunctions.rs:13:26
   |
13 | fn destructured_argument((a, b): (i32, i32)) {}
   |                          ^^^^^^

error: required arguments after an `Option<_>` argument are ambiguous
       = help: add a `#[pyo3(signature)]` annotation on this function to unambiguously specify the default values for all optional parameters
  --> tests/ui/invalid_pyfunctions.rs:16:63
   |
16 | fn function_with_required_after_option(_opt: Option<i32>, _x: i32) {}
   |                                                               ^^^

error: `cancel_handle` attribute only allowed with `async fn`
  --> tests/ui/invalid_pyfunctions.rs:18:14
   |
18 | #[pyfunction(cancel_handle = "cancel")]
   |              ^^^^^^^^^^^^^

error: missing cancel_handle argument
  --> tests/ui/invalid_pyfunctions.rs:21:30
   |
21 | #[pyfunction(cancel_handle = "cancel")]
   |                              ^^^^^^^^
